eventlog/eventlog.py

```diff
@@ -0,0 +1,882 @@
+            "EventNum": self.evidx,
+            "PCRIndex": self.evpcr,
+            "EventSize": self.evsize,
+            "DigestCount": len(self.digests),
+            "Digests": list(self.digests.values()),
+            "Event": self.evbuf[:1024].hex(),
+        }
+
+
+class ValidatedEvent(GenericEvent):
+    """
+    Events that can be validated by CONTENT_MATCHES_DIGEST
+    (TCG Guidance on Integrity Measurements and Event Log Processing, V1, Rev 0.118, 12/15/2021, Section 7.2.5.1)
+    EV_S_CRTM_VERSION EV_EFI_VARIABLE_DRIVER_CONFIG EV_SEPARATOR EV_EFI_GPT_EVENT EV_EFI_VARIABLE_BOOT
+    """
+
+    def validate(self) -> Tuple[bool, bool, str]:
+        for algid, refdigest in self.digests.items():
+            calchash1 = EfiEventDigest.hashalgmap[algid](self.evbuf).digest()
+            if refdigest != calchash1:
+                return False, False, str(self.evtype.name)
+        return False, True, ""
+
+
+class PostCodeEvent(GenericEvent):
+    """
+    EV_POST_CODE -- interpreted as a string or else as a blob base/length pair
+    """
+
+    def __init__(self, eventheader: Tuple, buffer: bytes, idx: int):
+        super().__init__(eventheader, buffer, idx)
+        if self.evsize == 16:
+            (self.blobBase, self.blobLength) = struct.unpack(
+                "<QQ", buffer[idx : idx + 16]
+            )
+        else:
+            self.blobBase = None
+            self.blobLength = None
+
+    def to_json(self) -> dict:
+        if self.blobBase is not None:
+            evt = {"BlobBase": self.blobBase, "BlobLength": self.blobLength}
+        else:
+            evt = self.evbuf.decode("utf-8")
+        return {**super().to_json(), "Event": evt}
+
+
+class FirmwareBlobEvent(GenericEvent):
+    """
+    Firmware blob measurement (old and new)
+    EV_EFI_PLATFORM_FIRMWARE_BLOB EV_EFI_PLATFORM_FIRMWARE_BLOB2
+    """
+
+    def __init__(self, eventheader: Tuple, buffer: bytes, idx: int):
+        super().__init__(eventheader, buffer, idx)
+        (self.base, self.length) = struct.unpack("<QQ", buffer[idx : idx + 16])
+
+    def to_json(self) -> dict:
+        return {
+            **super().to_json(),
+            "Event": {"BlobBase": self.base, "BlobLength": self.length},
+        }
+
+
+class EfiIPLEvent(GenericEvent):
+    """
+    EV_IPL: Used during initial program load.
+    ---
+    Differs from generic events in that the body of the
+    event is a zero terminated UTF-8 string describing
+    what is being loaded.
+
+    NOTE we choose to not carry the terminating zero as part of the JSON output.
+    """
+
+    def to_json(self) -> dict:
+        return {**super().to_json(), "Event": {"String": nullterm8(self.evbuf[:-1])}}
+
+
+class SpecIdEvent(GenericEvent):
+    """
+    This is the first event in the log, and it gets a lot of
+    special processing. It does not have a type in the enumeration
+    of event types.
+    """
+
+    def __init__(self, eventheader: Tuple, buffer: bytes, idx: int):
```

Comment:
Following `Tuple` approach, this (and the rest of signatures) needs to be also `Tuple[int, int, dict, int, int]`?